smolyay/surrogate.py

```diff
@@ -1,546 +1,742 @@
+        self._basis_sets = None
+        self._regularization = None
+        self._index_combinations = None
+        self._coefficients = None
+        self._integration_constant = 0
+        self._terms_constructed_cache = False
+
+        self._basis_sets = basis_sets
+        self.regularization = regularization
+
+    @property
+    def regularization(self):
+        """:class:RegularizationHelper: constant of the L2 and L1 term"""
+        return self._regularization
+
+    @regularization.setter
+    def regularization(self, value):
+        if value is not None and not isinstance(value, RegularizationMethod):
+            raise ValueError("Regression must be a RegularizationHelper")
+        if self.regularization != value:
+            self._regularization = value
+            self._valid_cache = False
+
+    @property
+    def basis_sets(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        return self._basis_sets
+
+    def predict(self, X):
         """Evaluate surrogate at a given input.
 
-        Parameter
-        ---------
-        x: list (for dimension > 1), or float (for dimension = 1)
-            Input.
+        Parameters
+        ----------
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated.
 
         Returns
         -------
-        float
+        ndarray of shape (n_samples,) or (n_samples, n_features)
             Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
         IndexError
-            Input of the surrogate must be of length of the
-            function's dimensionality.
+            Input must be 2D array with shape (n_samples, n_features).
         ValueError
             Input must lie in domain of surrogate.
+
         """
-        if self._coefficients is None:
-            raise ValueError('Function needs training!')
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-        
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
-        else:
-            oob = (numpy.any(x < self.domain[0]) or 
-                    numpy.any(x > self.domain[1]))
+        # validate inputs
+        if not self._valid_cache:
+            raise RuntimeError("Model must be fit!")
+        X = numpy.array(X, ndmin=2)
```

Comment:
Do you need to specify a `copy` argument here to avoid copying all the time? In NumPy 2, it should be `copy=None`. With older NumPy, it is different. We can require NumPy >= 2 since this is new software if we want.

Reply:
I just realized I've been developing on numpy version 1.24.3, so numpy < 2 is probably what a lot of people who aren't vigilante about updating packages are using.

We can use `X = numpy.asarray(X)` instead. The only scenario that passes the shape check with `X = numpy.array(X, ndmin=2)` but not `X = numpy.asarray(X)` is a 1D surrogate with only one point. We could turn X into a 2D array for that specific scenario.

Reply:
We will eventually set this up so it can be installed with pip, so we can force NumPy 2 at installation time. I would rather not have to maintain compatibility with both... we do it in relentless, and it is annoying.

Reply:
Upgraded my version and fixed some compatibility issues.